### ISSUE TITLE
Use of `prefix: /` in `packages.yaml` can cause traversal of entire file system

### ISSUE THREAD

**Issue by bcdarwin** [LLNL/spack#24418]
It appears that `prefix: /` in `packages.yaml` can cause Spack to traverse the entire file system.

For example, on an Ubuntu 20.04 system, `spack external find` nondeterministically locates `xz` as having prefix either `/` or `/usr` (though I don't think there's anything specific to `xz` here, just for minimality since `packages.yaml` seemingly must have at least one package).  The former case is problematic:

### Steps to reproduce the issue

```console
$ git clone https://github.com/spack/spack.git
$ cd spack
$ spack compiler find --scope=site
$ spack external find xz --scope=site
# assume spack chooses `/`,
# or edit `etc/spack/packages.yaml` as follows:
packages:
  xz:
    externals:
    - spec: xz@5.2.4
      prefix: /
$ spack install python@3.8.10
 ...
==> Ran patch() for python
```

At this point the build appears to hang, and `strace` reveals calls to `lstat` on files in my `/projects` directory, which is immense.

### Information on your system

```
* **Spack:** 0.16.2-3199-383d4cc84c
* **Python:** 3.8.5
* **Platform:** linux-ubuntu20.04-broadwell
* **Concretizer:** original
```

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [ ] I have searched the issues of this repo and believe this is not a duplicate
- [ ] I have run the failing commands in debug mode and reported the output

**Comment by adamjstewart** [comment]
Whenever a Spack package like `python` depends on another package like `xz`, Spack queries the `xz` package to find out where it's libraries are installed. In the case of `xz`, it does this by searching for `liblzma` in the following locations:

1. `prefix/lib`
2. `prefix/lib64`
3. `prefix` (recursively)

So if the library isn't found in the first two locations, we resort to a recursive search. This seems to be what you're experiencing.

Basically, I wouldn't consider the fact that Spack is searching the entire filesystem to be a bug (it has to find the library somehow), but I would consider the fact that `spack external find` returns `/` when there is no library in `/lib` or `/lib64` to be a bug.

Is your `/` directory and/or `/bin/xz` file a symlink to the actual installation in `/usr`? Do you have a `liblzma.so` file somewhere in `/usr/lib` or `/usr/lib64`?

**Comment by bcdarwin** [comment]
In Ubuntu 20.04, due to upstream changes in Debian, /bin, /sbin, /lib, /lib64 are symlinks to usr/bin, usr/sbin, usr/lib, and usr/lib64, respectively, so the libraries detected in each case should be the same.

**Comment by bcdarwin** [comment]
[Of course this means there's an easy workaround in my situation.]

**Comment by adamjstewart** [comment]
In that case I think we can tell `spack external find` to resolve symlinks and use the full path to the executable. Let me see if I can figure out where to do that.

**Comment by bcdarwin** [comment]
So with your PR, `spack external find` produces a `packages.yaml` with all prefixes being `/usr` instead of `/`.  In a sense this resolves this likely resolves the issue as presenting on most Ubuntu 20.04 systems.  However, I am not sure this is the most widely applicable resolution, since the original issue could still reoccur for various reasons, such as a change in system packages, manual editing of packages.yaml, or packages only present in `/` but not `/usr` which somehow trigger the expensive directory traversal.  I suspect that the real issue is trying to reconstruct paths from a `prefix` rather than storing slightly more information.  But, of course, I'm not too familiar with Spack's logic or architecture so you may have reasons not to go down this route.

**Comment by adamjstewart** [comment]
In the long run, we want to be able to store more information in `packages.yaml` like you suggest. For example, there are systems where executables are in `/usr/bin` but libraries are in `/usr/lib/x86_64_linux_gnu` with no symlinks. However, even then we'll still need to be able to recursively search for libraries in order to auto-populate `packages.yaml` for users. Of course, if users edit their own file, we wouldn't have encountered this issue in the first place.

Another idea is to limit the recursion depth to 1 or 2 subdirectories. However, this won't work for Python packages where libraries live in `lib/pythonX.Y/site-packages/foo/**`.

**Comment by haampie** [comment]
> spack external find nondeterministically locates xz as having prefix either `/` or `/usr`

This detail was fixed by #46957.

That does not make the issue go away in general, given that packages do `spec["pkg"].libs` and `spec["pkg"].headers`, triggering an exhaustive search.

**Comment by haampie** [comment]
My suggestion is to just recurse in `<prefix>/{lib,lib64}` (and probably also `<prefix>/bin` due to windows). Limiting depth is then slightly less relevant.

**Comment by haampie** [comment]
@johnwparent pointed out on windows things may be placed directly under `prefix/`.

So, a possible way to speed this up would be a combined

```
find(prefix, recursive=False) + find([prefix.bin, prefix.lib64, prefix.lib], recursive=True, max_depth=3)
```

which would still search @adamjstewart's favorite `site-packages/*` dirs.

It's not helping with any other issues like spack searching the wrong multiarch dir on debian

**Comment by haampie** [comment]
After some more digging, there are still at least two things to consider. In short `find_libraries` is already an opinionated function which I think is fine.

1. `find_libraries` currently *has* search heuristics or shortcuts. It does *not* do exhaustive search with `recursive=True` if matching libraries are found first in `root/lib` or `root/lib64`. If not, slow recursive search of  `root/` is performed. That slow path is however *always* hit on Debian with externals in `/usr` due to multi-arch dirs `/usr/lib/<triplet>`.

2. We should decide whether `find_libraries(prefix.lib)` in `package.py` is a bug or not. Some packages do this likely to speed up search, but libraries may actually be on lib64 depending on the distro. Some other packages do `libs = find_libraries(prefix.lib) + find_libraries(prefix.lib64)`, which is OK-ish except for Windows.

Maybe the least breaking way to deal with (2) is to improve the heuristic to take into account whether the input directory itself is already a typical library dir and use the same heuristics we do for (1) on those dirs:

```python
def find_libraries(root, max_depth=3):
    if os.path.basename(root) in ("lib", "lib64", "bin"):
        files = find(root, recursive=True, max_depth=default_max_depth)
    else:
        files = find([root / "lib", root / "lib64", root / "bin"], max_depth=default_max_depth) + find(root, recursive=False)
    ...
```

In this implementation the performance is better in the unhappy path due to `max_depth` on library subdirs and non-recursive search on the root, BUT it's slower in the typical happy path *without* externals, since libs are then typically directly in `root/lib` and `root/lib64`.

Alternative implementation that hopefully is good enough for the happy and unhappy code paths:

```python

def find_libraries(root, max_depth=3):
    if os.path.basename(root) in ("lib", "lib64", "bin"):
        files = find(root, recursive=True, max_depth=max_depth)
    else:
        # do not recurse in the usual suspects so we can early return
        files = find([root, root / "lib", root / "lib64", root / "bin"], recursive=False)
        if not files:
            # only recurse into typical lib dirs, but exclude root
            files = find([root / "lib", root / "lib64", root / "bin"], max_depth=max_depth)
```

**Comment by alalazo** [comment]
> Alternative implementation that hopefully is good enough for the happy and unhappy code paths:

:point_up: This heuristic seems :100: for v0.23 We can think of bigger changes, maybe involving externals and their definitions, after the release.

**Comment by haampie** [comment]
The only downside of that implementation is that it's still potentially slow for externals in /usr on Debian if max_depth is say 3.

Another search method is iterative deepening, which we in fact already use for `find_first`. 

```python
    for depth in range(0, max_depth):
        files = find(..., min_depth=depth, max_depth=depth)  # min_depth is to avoid running regexes on low-depth files already seen
        if files:
            break
```

which could have a slightly smarter implementation whether BFS or DFS (but neither are entirely satisfactory since I think BFS will have redundant stat calls on leaf nodes, and stateless DFS will have repeated stat calls on internal nodes).
